### PR TITLE
Add copyto to facilitate broadcasting

### PR DESCRIPTION
**Pull request by jishnub**
This PR adds a custom broadcast style for `AbstractKroneckerProducts` which uses the efficient `collect!` wherever possible to evaluate broadcast operations without indexing into the Kronecker product (which is slow). This often leads to significant speed-ups.

On master
```julia
julia> K1 = kronecker(ones(30, 30), ones(30, 30));

julia> K2 = kronecker(ones(30, 30), ones(30, 30));

julia> @btime $K1 + $K2;
  56.020 ms (2 allocations: 6.18 MiB)

julia> @btime $K1 .+ $K2;
  55.951 ms (2 allocations: 6.18 MiB)

julia> @btime -$K1;
  28.102 ms (2 allocations: 6.18 MiB)

julia> @btime copy($K1);
  27.994 ms (2 allocations: 6.18 MiB)
```

The timing is dominated entirely by indexing

```julia
julia> @btime $K1[1,1];
  34.343 ns (0 allocations: 0 bytes)

julia> 34.343 * 1e-6 * length(K1) # time in ms to loop over the array elementwise
27.81783
```

After this PR

```julia
julia> @btime $K1 + $K2;
  1.503 ms (2 allocations: 6.18 MiB)

julia> @btime $K1 .+ $K2;
  1.491 ms (2 allocations: 6.18 MiB)

julia> @btime -$K1;
  1.047 ms (2 allocations: 6.18 MiB)

julia> @btime copy($K1);
  735.405 μs (2 allocations: 6.18 MiB)
```

The speed of the sum is comparable to that of dense arrays now:

```julia
julia> @btime $(collect(K1)) + $(collect(K2));
  1.714 ms (2 allocations: 6.18 MiB)
```

### CODE REVIEW

File: src/kroneckerpowers.jl [hunk 1]
```diff
@@ -15,10 +15,10 @@ K = A ⊗ A ⊗ ... ⊗ A.
 """
 struct KroneckerPower{T<:Any,TA<:AbstractMatrix{T}, N} <: AbstractKroneckerProduct{T}
    A::TA
-   pow::Integer
+   pow::Int
    function KroneckerPower(A::AbstractMatrix{T}, pow::Integer) where {T}
       @assert pow ≥ 2 "KroneckerPower only makes sense for powers greater than 1"
-      return new{eltype(A), typeof(A), pow}(A, pow)
+      return new{eltype(A), typeof(A), Int(pow)}(A, Int(pow))
```
Comment (jishnub) [inline src/kroneckerpowers.jl:21]: Since `pow` is a type-parameter, it must be a bitstype object, so we may cast it to `Int`.

File: src/base.jl [hunk 2]
```diff
@@ -91,7 +91,7 @@ function getindex(K::AbstractKroneckerProduct, i1::Integer, i2::Integer)
     A, B = getmatrices(K)
     m, n = size(A)
     k, l = size(B)
-    return A[cld(i1, k), cld(i2, l)] * B[(i1 - 1) % k + 1, (i2 - 1) % l + 1]
+    return (A[cld(i1, k), cld(i2, l)]::eltype(A)) * (B[(i1 - 1) % k + 1, (i2 - 1) % l + 1]::eltype(B))
```
Comment (jishnub) [inline src/base.jl:94]: type-assertions here help if the individual `getindex` fails to infer the type, eg. in `KroneckerPower`

**Comment by MichielStock** [comment]
Wow, very nice work! Great job!

**Comment by jishnub** [comment]
Could you tag a new release? Thanks!

**Comment by MichielStock** [comment]
Requested and scheduled!

**Pull request merged**